core/base-service/base.js

```diff
@@ -214,20 +214,43 @@ class BaseService {
     return result
   }
 
-  constructor({ sendAndCacheRequest, authHelper }, { handleInternalErrors }) {
+  constructor(
+    { sendAndCacheRequest, authHelper, metricHelper },
+    { handleInternalErrors }
+  ) {
     this._requestFetcher = sendAndCacheRequest
     this.authHelper = authHelper
     this._handleInternalErrors = handleInternalErrors
+    this._metricHelper = metricHelper
   }
 
   async _request({ url, options = {}, errorMessages = {} }) {
     const logTrace = (...args) => trace.logTrace('fetch', ...args)
     logTrace(emojic.bowAndArrow, 'Request', url, '\n', options)
     const { res, buffer } = await this._requestFetcher(url, options)
+    await this._meterResponse(res, buffer)
```

Comment:
Is it necessary to `await` this one or is it possible to process the response/return the badge to the user and send the response size data to prometheus in parallel/afterwards?

Reply:
Thanks for this accurate comment! 
`await` is it not necessary at all. Prometheus uses pull model over HTTP and we expose our metrics at `/metrics`. We store metrics in memory and we do not send them anywhere (no network calls). This particular metric is currently a histogram and collecting data in histograms is not a CPU intensive operation (in the histogram "Observations are very cheap as they only need to increment counters."). But this may change in the future - we may change histogram to something else or event replace Prometheus with an other monitoring solution which can do network calls. 
Therefore we should gather metrics in the background and return a response to the user  without a delay. I will remove `await`. 

After removing `await` `service response size metric is optional` test is passing, but it should wait for `expect` to be true. Do you have a suggestion how we can test it? My proposition is to use https://github.com/TheBrainFamily/wait-for-expect.

Reply:
> After removing await service response size metric is optional test is passing, but it should wait for expect to be true

I'm not sure I follow, could you elaborate a bit more? Are you saying that the test is passing, but that the validation/assertion statement is basically invalid (because the test is skipping it)? 

Also am I correct that this is the test you are referring to?
https://github.com/badges/shields/pull/4434/files#diff-0567f28266920ae49126b66b59269cdfR516-R539

Reply:
That's right. https://github.com/badges/shields/pull/4434/files#diff-0567f28266920ae49126b66b59269cdfR516-R539 is testing `_request` method. In the test we are waiting for the result of the `await serviceInstance._request({ url, options })`, but `_request` is collecting metrics by async method `_meterResponse` in parallel. In the test we are not waiting for `_meterResponse` to finish. Currently metrics are collected in memory and test is passing. But in the future this can change. We can simulate it by:
```diff
@@ -243,4 +243,5 @@ class BaseService {
 
   async _meterResponse(res, buffer) {
+    await new Promise((resolve) => setTimeout(resolve, 1))
     if (
       this._metricHelper &&
```

Reply:
Gotcha, thanks for the clarification. I hadn't looked closely enough at the innards of `_meterResponse`. 

By any chance do we have any profiling/measurement available? I'd venture to say that the easiest thing for now would be to keep that `await` in there, as I'm guessing there's only a very very small amount of time needed right now to cache/queue the metric in the current setup. 

Similar to the other comments here, this is something that could be revisited if there's any refactoring here.

Reply:
I restored this `await`. Saving this data should take negligible amount of time.